ports/nrf/mphalport.c

```diff
@@ -35,6 +35,71 @@
 #include "nrfx_errors.h"
 #include "nrfx_config.h"
 
+#if MICROPY_PY_TIME_TICKS
+#include "nrfx_rtc.h"
+#endif
+
+
+#if MICROPY_PY_TIME_TICKS
+// use RTC1 for time ticks generation
+// with 32kHz tick resolution and overflow
+// handling in irq
+
+nrfx_rtc_t rtc1 = NRFX_RTC_INSTANCE(1);
+
+volatile mp_uint_t rtc_overflows = 0;
+
+const nrfx_rtc_config_t rtc_config_time_ticks = {
+    .prescaler    = 0,
+    .reliable     = 0,
+    .tick_latency = 0,
+    #ifdef NRF51
+    .interrupt_priority = 1,
+    #else
+    .interrupt_priority = 3,
+    #endif
+};
+
+STATIC void rtc_irq_time(nrfx_rtc_int_type_t event) {
+    if (event == NRFX_RTC_INT_OVERFLOW) {
+        rtc_overflows += 1;
+    }
+}
+
+void rtc1_init_time_ticks(void) {
+    nrfx_rtc_init(&rtc1, &rtc_config_time_ticks, rtc_irq_time);
+    nrfx_rtc_overflow_enable(&rtc1, true);
+    nrfx_rtc_enable(&rtc1);
+}
+
+mp_uint_t mp_hal_ticks_ms(void) {
+    // XXX check overflow bit here? (if rtc1.p_reg->EVENTS_OVRFLW ..)
+    // note that COUNTER * 1000 / 32768 would overflow during calculation, so use
+    // the less obvious * 125 / 4096 calculation (overflow secure)
+    return (rtc_overflows << 9) * 1000 + ((mp_uint_t)rtc1.p_reg->COUNTER * 125 / 4096);
+}
+
+mp_uint_t mp_hal_ticks_us(void) {
+    // make the best out of the 32kHz tick resolution (= 30.57usec / tick)
+    // to scale we could use same approach as above and use COUNTER * 125000 / 4096
+    // but that will overflow for 32bit multiplications.
+    //
+    // Since this function is likely to be called in a poll loop it must
+    // be fast and casting to 64 bits introduces calculation timing within that loop,
+    // causing timing inaccuracy.
+    //
+    // An easier solution is to use COUNTER * 32, but introduces an error
+    // of app. 5%. However lets consider this as a tradeoff since the tick's rate
+    // is also very rough and not really suitable for low value usec timings.
+    return (mp_uint_t)rtc1.p_reg->COUNTER * 32;
```

Comment:
It can use `COUNTER * 214 / 7` which gives an error of 0.18% and will not overflow 32 bits.

Reply:
Ok -  I was thinking about a simple division using a 2**x divisor to minimize overhead. But actually that's not really needed. Will adapt this.

Reply:
I understand.  But I think the overhead of `/7` is small enough (the `*214` should be very efficient even without hardware mult).

Reply:
agree